homeassistant/components/tuya/config_flow.py

```diff
@@ -108,6 +108,7 @@ def _try_connect(self):
         except TuyaAPIException:
             return RESULT_AUTH_FAILED
 
+        self.hass.data[DOMAIN] = {TUYA_DATA: tuya}
```

Comment:
How about doing this if tuya.init returns true?  If false, reset the store so that auth and connection will be started from scratch.

```
if tuya.init(...): 
    self.haas.data[DOMAIN] = {TUYA_DATA: tuya}
else:
    self.haas.data[DOMAIN] = {TUYA_DATA: None}
```

Reply:
We don't allow the config flow to set instances in hass.data. The config flow should be isolated.

Reply:
@MartinHjelmare sounds like this is a "best practice" restriction? The code did run when I tested it. I'm happy to change this around but I'm not sure if there is any other global state, or way to forward config objects to the actual setup of the component.

Reply:
Maybe there is some other established pattern for handling heavy rate limiting like this already. Open to any suggestions.

Reply:
Now we have an integration that say OK during configuration and than is not created. If this solution is not applicable, probably is necessary to find an alternative that delay component creation and inform the user after integration setup.

Reply:
When the library has been updated to raise the specific exception we can catch that in `async_setup_entry` and raise `ConfigEntryNotReady`. That will retry the setup.

Reply:
So we know that cannot be done. At this point only solution is wait for upstream PR and create new PR here when ready.

Reply:
That's still a pretty miserable experience though. Even if we catch and wait, the UI will look as though everything froze. Hopefully tuya gets back to me and we can make a change to the API to better handle this. In the meantime I guess this PR can be declined.

Reply:
The UI won't look frozen. This problem only happens when first creating the integration. So nothing from this integration will show up until the entry setup finishes after the retry is successful.

Reply:
I'll try this again shortly, maybe I did something wrong before